feat(docs): Mount External Storage (#4673)
daytonaio/daytona: apps/docs/src/content/docs/en/mount-external-storage.mdx

---
title: Mount External Storage
description: Mount external object storage (Amazon S3, Cloudflare R2, Google Cloud Storage, Azure Blob) and filesystems into Daytona sandboxes using FUSE.
---

import { TabItem, Tabs } from '@astrojs/starlight/components'

> Mount object storage (Amazon S3, Cloudflare R2, Google Cloud Storage, Azure Blob) and filesystems like MesaFS into a Daytona sandbox as a regular directory. The sandbox reads and writes the bucket like local files, so existing tools, scripts, and agents work without changes. This is useful for bringing in datasets, model weights, or build artifacts that already live in your own cloud account.

Reviewer: ```suggestion
Mount object storage (Amazon S3, Cloudflare R2, Google Cloud Storage, Azure Blob) and filesystems like MesaFS into a Daytona sandbox as a regular directory. The sandbox reads from and writes to the bucket as if it were a local directory, so existing tools, scripts, and agents work without changes. This is useful for bringing in datasets, model weights, or build artifacts that already live in your own cloud account.
```

> External storage mounts and [Daytona Volumes](/docs/en/volumes/) are complementary FUSE-based mechanisms — both expose remote object storage as a regular sandbox directory, both can be shared across sandboxes, and both persist beyond any individual sandbox's lifetime. The main distinction is **where the data physically lives**: Daytona Volumes are hosted on Daytona's own S3-compatible object store, while external mounts attach a bucket or filesystem hosted on another provider (Amazon S3, Cloudflare R2, GCS, Azure Blob, MesaFS).

Reviewer: ```suggestion
External storage mounts and [Daytona Volumes](/docs/en/volumes/) are complementary FUSE-based mechanisms — both expose remote object storage as a regular sandbox directory, both can be shared across sandboxes, and both persist beyond any individual sandbox's lifetime. The main distinction is **where the data physically lives**: Daytona Volumes are hosted on Daytona's own S3-compatible object store, while external mounts connect to a bucket or filesystem hosted on another provider (Amazon S3, Cloudflare R2, GCS, Azure Blob, MesaFS).
```

## How it works

> External storage is mounted using FUSE. Daytona supports two approaches and you can pick whichever fits your workflow:

Reviewer: ```suggestion
External storage is mounted using FUSE. Daytona supports two approaches, and you can pick whichever fits your workflow:
```

- **Pre-baked snapshot** — build a [snapshot](/docs/en/snapshots/) once with the FUSE tool (`mount-s3`, `gcsfuse`, `blobfuse2`) baked in, then launch every sandbox from that snapshot. Cold starts are fast and predictable. Best for production.
- **Runtime install** — launch a default sandbox and `apt-get install` the FUSE tool when the sandbox starts. Adds time to sandbox startup, but you don't manage snapshots. Best for quick experiments.

Both approaches end with the same mount command and the same usage — the only difference is when the FUSE tool gets installed.

> Each provider section below shows both approaches.

Reviewer 2: Suggestion to remove the "how it works" heading and merge the section with the above description paragraphs.

## Prerequisites

- The [Daytona SDK](/docs/en/getting-started/) installed for the language you're using

> - Provider credentials (AWS keys, GCS service account JSON, Azure account key, etc.)

Reviewer 2: I'd remove the first bullet point as it seems redundant to mention you need Daytona, and then move the second bullet under each of the providers separately (optionally also explain where/how to add them).

Author: Like the idea, applied

## Mount an Amazon S3 bucket

Mount an S3 bucket using [Mountpoint for Amazon S3 ↗](https://github.com/awslabs/mountpoint-s3) — AWS's official FUSE client, optimized for high throughput on S3.

### Pre-baked snapshot

Build a snapshot once with `mount-s3` installed, then launch sandboxes from it.

> #### Build the snapshot

Reviewer 2: ```suggestion
#### Build a snapshot
```

Author: Applied everyhwere

                    + "&& sudo apt-get install -y --no-install-recommends libfuse2 ca-certificates wget");
            sandbox.getProcess().executeCommand(
                "arch=\"$(dpkg --print-architecture | sed s/amd64/x86_64/)\" "
                    + "&& wget -O /tmp/mount-s3.deb "
                    + "\"https://s3.amazonaws.com/mountpoint-s3-release/latest/${arch}/mount-s3.deb\" "
                    + "&& sudo apt-get install -y /tmp/mount-s3.deb");

            // Mount and use
            String mountPath = "/home/daytona/s3";
            sandbox.getProcess().executeCommand(
                "mkdir -p " + mountPath + " && mount-s3 my-bucket " + mountPath);
            ExecuteResponse response = sandbox.getProcess().executeCommand("ls " + mountPath);
            System.out.println(response.getResult());
        }
    }
}
```

</TabItem>
</Tabs>

## Mount a Cloudflare R2 bucket

Cloudflare R2 is S3-compatible, so the same `mount-s3` tool works. Pass an explicit `--endpoint-url` pointing at your R2 account.

### Pre-baked snapshot

R2 reuses the `fuse-s3` snapshot built in the [S3 section](#build-the-snapshot). If you already have it, skip straight to launch and mount.

> 

Reviewer 2: I'd repeat the missing section here to have it accessible regardless.

Author: Added

                    + "&& wget -O /tmp/mount-s3.deb "
                    + "\"https://s3.amazonaws.com/mountpoint-s3-release/latest/${arch}/mount-s3.deb\" "
                    + "&& sudo apt-get install -y /tmp/mount-s3.deb");

            // Mount with R2 endpoint
            String mountPath = "/home/daytona/r2";
            sandbox.getProcess().executeCommand(
                "mkdir -p " + mountPath + " && "
                    + "mount-s3 --endpoint-url https://" + accountId + ".r2.cloudflarestorage.com "
                    + "my-r2-bucket " + mountPath);

            ExecuteResponse response = sandbox.getProcess().executeCommand("ls " + mountPath);
            System.out.println(response.getResult());
        }
    }
}
```

</TabItem>
</Tabs>

## Mount a Google Cloud Storage bucket

Mount a GCS bucket using [gcsfuse ↗](https://github.com/GoogleCloudPlatform/gcsfuse) — Google's official FUSE client.

:::note
Daytona's base image is Debian Trixie, but Google has not published a Trixie-specific gcsfuse repository yet. The `gcsfuse-bookworm` repo packages run cleanly on Trixie, so we use them in the snippets below.
:::

### Pre-baked snapshot

> #### Build the snapshot

Reviewer 2: ```suggestion
#### Build a snapshot
```

}
```

</TabItem>
<TabItem label="Java" icon="seti:java">

```java
import io.daytona.sdk.Daytona;
import io.daytona.sdk.Image;

public class App {
    public static void main(String[] args) {
        try (Daytona daytona = new Daytona()) {
            Image image = Image.base("daytonaio/sandbox")
                .runCommands(
                    "sudo apt-get update "
                        + "&& sudo apt-get install -y --no-install-recommends ca-certificates curl gnupg",
                    "sudo mkdir -p /etc/apt/keyrings "
                        + "&& curl -fsSL https://packages.cloud.google.com/apt/doc/apt-key.gpg "
                        + "| sudo gpg --dearmor -o /etc/apt/keyrings/gcsfuse.gpg",
                    "echo \"deb [signed-by=/etc/apt/keyrings/gcsfuse.gpg] "
                        + "https://packages.cloud.google.com/apt gcsfuse-bookworm main\" "
                        + "| sudo tee /etc/apt/sources.list.d/gcsfuse.list",
                    "sudo apt-get update && sudo apt-get install -y gcsfuse"
                );

            daytona.snapshot().create("fuse-gcs", image, System.out::println);
        }
    }
}
```

</TabItem>
</Tabs>

#### Launch and mount

> GCS authenticates with a service account JSON key. Upload it into the sandbox via `sandbox.fs` and point `gcsfuse` at it with `--key-file`.

Reviewer: ```suggestion
`gcsfuse` authenticates to GCS with a service account JSON key. Upload it into the sandbox via `sandbox.fs` and point `gcsfuse` at it with `--key-file`.
```

            String mountPath = "/home/daytona/gcs";
            String keyPath = "/home/daytona/.gcs-key.json";
            sandbox.fs.uploadFile(serviceAccountKey, keyPath);
            sandbox.getProcess().executeCommand("chmod 600 " + keyPath);
            sandbox.getProcess().executeCommand("mkdir -p " + mountPath);
            sandbox.getProcess().executeCommand(
                "gcsfuse --key-file=" + keyPath + " my-gcs-bucket " + mountPath);

            ExecuteResponse response = sandbox.getProcess().executeCommand("ls " + mountPath);
            System.out.println(response.getResult());
        }
    }
}
```

</TabItem>
</Tabs>

## Mount an Azure Blob container

Mount an Azure Blob container using [blobfuse2 ↗](https://github.com/Azure/azure-storage-fuse) — Microsoft's official FUSE client.

:::caution
Daytona's base image is Debian Trixie. Microsoft's `blobfuse2` Bookworm binary links against `libfuse3.so.3`, but Trixie's `fuse3` package bumped the SONAME to `.4`, so `libfuse3.so.3` is missing on disk. The snapshot recipe below creates a compatibility symlink. Without it, `blobfuse2` fails with `libfuse3.so.3: cannot open shared object file`.
:::

### Pre-baked snapshot

> #### Build the snapshot

Reviewer 2: ```suggestion
#### Build a snapshot
```

                        + "signed-by=/etc/apt/trusted.gpg.d/microsoft.gpg] "
                        + "https://packages.microsoft.com/debian/12/prod bookworm main\" "
                        + "| sudo tee /etc/apt/sources.list.d/microsoft-prod.list",
                    "sudo apt-get update && sudo apt-get install -y blobfuse2 fuse3",
                    // libfuse3.so.3 compat symlink for Trixie
                    "src=$(find /usr/lib /lib -name \"libfuse3.so.3.*\" -type f 2>/dev/null "
                        + "| sort -V | tail -1) "
                        + "&& sudo ln -sfn \"$src\" \"$(dirname \"$src\")/libfuse3.so.3\" "
                        + "&& sudo ldconfig",
                    "sudo touch /etc/fuse.conf "
                        + "&& grep -qxF \"user_allow_other\" /etc/fuse.conf "
                        + "|| echo \"user_allow_other\" | sudo tee -a /etc/fuse.conf"
                );

            daytona.snapshot().create("fuse-azure", image, System.out::println);
        }
    }
}
```

</TabItem>
</Tabs>

#### Launch and mount

`blobfuse2` reads its configuration from a YAML file. Build it with your account credentials and upload it into the sandbox.

> The YAML below tells `blobfuse2` three things: **where to connect** (the `azstorage:` block — your storage account, the container you want to mount, the endpoint URL, and the auth method), **what to enable** (the `components:` list — the FUSE interface itself, a content cache, a metadata cache, and the Azure backend), and **how to log**. The cache components use sensible defaults when listed without their own top-level config blocks; add explicit `block_cache:` / `attr_cache:` blocks later if you need to tune cache sizes or timeouts. Note that in Azure terminology a "container" is the equivalent of an S3 bucket — it's specified inside the YAML rather than passed as a command-line argument.

Reviewer: ```suggestion
The YAML below tells `blobfuse2` three things: **where to connect** (the `azstorage:` block — your storage account, the container you want to mount, the endpoint URL, and the auth method), **what to enable** (the `components:` list — the FUSE interface itself, a content cache, a metadata cache, and the Azure backend), and **how to log**. The cache components use sensible defaults when listed without their own top-level config blocks; add explicit `block_cache:` / `attr_cache:` blocks later if you need to tune cache sizes or timeouts. Note that in Azure terminology, a "container" is the equivalent of an S3 bucket — it's specified inside the YAML rather than passed as a command-line argument.
```

            String config = "allow-other: true\n"
                + "components:\n"
                + "  - libfuse\n"
                + "  - block_cache\n"
                + "  - attr_cache\n"
                + "  - azstorage\n"
                + "azstorage:\n"
                + "  type: block\n"
                + "  account-name: " + account + "\n"
                + "  container: " + container + "\n"
                + "  endpoint: https://" + account + ".blob.core.windows.net\n"
                + "  auth-type: key\n"
                + "  account-key: " + accountKey + "\n";

            sandbox.fs.uploadFile(config.getBytes(StandardCharsets.UTF_8), configPath);
            sandbox.getProcess().executeCommand("chmod 600 " + configPath);
            sandbox.getProcess().executeCommand("mkdir -p " + mountPath);
            sandbox.getProcess().executeCommand(
                "blobfuse2 mount --config-file=" + configPath + " " + mountPath);

            ExecuteResponse response = sandbox.getProcess().executeCommand("ls " + mountPath);
            System.out.println(response.getResult());
        }
    }
}
```

</TabItem>
</Tabs>

## Mount a MesaFS filesystem

> [MesaFS ↗](https://mesa.dev) is an agent-native versioned filesystem from Mesa, purpose-built for the same workloads Daytona sandboxes run — parallel agent swarms, shared working memory, structured artifacts, and long-lived state across runs. Unlike the cloud-bucket providers above, you mount a Mesa **repository**: a Git-compatible versioned working directory with sub-50ms reads/writes, instant fork/branch operations, and unlimited concurrent writers.

Reviewer: ```suggestion
[MesaFS ↗](https://mesa.dev) is an agent-native versioned filesystem from Mesa, purpose-built for the same workloads Daytona sandboxes run — parallel agent swarms, shared working memory, structured artifacts, and long-lived state across runs. With MesaFS, instead of mounting a cloud bucket, you mount a Mesa **repository**: a Git-compatible versioned working directory with sub-50ms reads/writes, instant fork/branch operations, and unlimited concurrent writers.
```

> The Mesa setup follows the same shape as the bucket providers but uses the Mesa CLI rather than a FUSE-specific tool: install the CLI in your sandbox, authenticate with your API key, and run `mesa mount --daemonize` to mount your repos at `/home/daytona/mesa/mnt/<org>/<repo>`.

Reviewer: ```suggestion
The Mesa setup follows the same pattern as the bucket providers but uses the Mesa CLI rather than a FUSE-specific tool: install the CLI in your sandbox, authenticate with your API key, and run `mesa mount --daemonize` to mount your repos at `/home/daytona/mesa/mnt/<org>/<repo>`.
```

### Pre-baked snapshot

> #### Build the snapshot

Reviewer 2: ```suggestion
#### Build a snapshot
```

<Tabs syncKey="language">
<TabItem label="Python" icon="seti:python">

```python
from daytona import CreateSnapshotParams, Daytona, Image

daytona = Daytona()

image = (
  Image.base("daytonaio/sandbox")
  .run_commands(
    "sudo apt-get update "
    "&& sudo apt-get install -y --no-install-recommends "
    "ca-certificates fuse3 libssl3 openssl "
    "&& sudo update-ca-certificates",
    "curl -fsSL https://mesa.dev/install.sh | sh "
    "&& sudo ln -sf $HOME/.mesa/bin/mesa /usr/local/bin/mesa",
    "sudo sed -i 's/^#user_allow_other/user_allow_other/' /etc/fuse.conf",
  )

> )

Reviewer: @OliverGilan, I think this could be simpler. Can you please confirm:

1. I only believe we need the `user_allow_other` change
2. I think this is only relevant `curl -fsSL https://mesa.dev/install.sh | sh `
3. The `ln` feels like an awful hack -- symlinking a user-installed binary into `/usr/local/bin` is a bad idea in general, for a multitude of reasons, and I believe that editing `PATH` is better. Perhaps we could make the `install.sh` script on our end support root installs.

Author: Thanks @markovejnovic — confirmed all three. Simplified to just `curl install.sh | sh` + `user_allow_other`.

)
```

</TabItem>
<TabItem label="TypeScript" icon="seti:typescript">

```typescript
import { Daytona, Image } from '@daytona/sdk'

const daytona = new Daytona()

const image = Image.base('daytonaio/sandbox').runCommands(
  'sudo apt-get update ' +
    '&& sudo apt-get install -y --no-install-recommends ' +
    'ca-certificates fuse3 libssl3 openssl ' +
    '&& sudo update-ca-certificates',
  'curl -fsSL https://mesa.dev/install.sh | sh ' +
    '&& sudo ln -sf $HOME/.mesa/bin/mesa /usr/local/bin/mesa',
  "sudo sed -i 's/^#user_allow_other/user_allow_other/' /etc/fuse.conf",
)

await daytona.snapshot.create(
  { name: 'fuse-mesa', image },
  { onLogs: console.log },
)
```

</TabItem>
<TabItem label="Ruby" icon="seti:ruby">

```ruby
require 'daytona'

daytona = Daytona::Daytona.new

image = Daytona::Image
  .base('daytonaio/sandbox')
  .run_commands(
    'sudo apt-get update ' \
    '&& sudo apt-get install -y --no-install-recommends ' \
    'ca-certificates fuse3 libssl3 openssl ' \
    '&& sudo update-ca-certificates',
    'curl -fsSL https://mesa.dev/install.sh | sh ' \
    '&& sudo ln -sf $HOME/.mesa/bin/mesa /usr/local/bin/mesa',

>     "sudo sed -i 's/^#user_allow_other/user_allow_other/' /etc/fuse.conf"

Reviewer: @OliverGilan, I think this could be simpler. Can you please confirm:

1. I only believe we need the `user_allow_other` change
2. I think this is only relevant `curl -fsSL https://mesa.dev/install.sh | sh `
3. The `ln` feels like an awful hack -- symlinking a user-installed binary into `/usr/local/bin` is a bad idea in general, for a multitude of reasons, and I believe that editing `PATH` is better. Perhaps we could make the `install.sh` script on our end support root installs.

  )

daytona.snapshot.create(
  Daytona::CreateSnapshotParams.new(name: 'fuse-mesa', image: image),
  on_logs: proc { |chunk| print(chunk) }
)
```

</TabItem>
<TabItem label="Go" icon="seti:go">

```go
import (
	"context"
	"fmt"
	"log"

	"github.com/daytonaio/daytona/libs/sdk-go/pkg/daytona"
	"github.com/daytonaio/daytona/libs/sdk-go/pkg/types"
)

ctx := context.Background()
client, err := daytona.NewClient()
if err != nil {
	log.Fatal(err)
}

image := daytona.Base("daytonaio/sandbox").
	Run("sudo apt-get update && sudo apt-get install -y --no-install-recommends " +
		"ca-certificates fuse3 libssl3 openssl && sudo update-ca-certificates").
	Run("curl -fsSL https://mesa.dev/install.sh | sh && " +
		"sudo ln -sf $HOME/.mesa/bin/mesa /usr/local/bin/mesa").
	Run(`sudo sed -i 's/^#user_allow_other/user_allow_other/' /etc/fuse.conf`)

_, logChan, err := client.Snapshot.Create(ctx, &types.CreateSnapshotParams{
	Name:  "fuse-mesa",
	Image: image,

> })

Reviewer: @OliverGilan, I think this could be simpler. Can you please confirm:

1. I only believe we need the `user_allow_other` change
2. I think this is only relevant `curl -fsSL https://mesa.dev/install.sh | sh `
3. The `ln` feels like an awful hack -- symlinking a user-installed binary into `/usr/local/bin` is a bad idea in general, for a multitude of reasons, and I believe that editing `PATH` is better. Perhaps we could make the `install.sh` script on our end support root installs.

if err != nil {
	log.Fatal(err)
}
for line := range logChan {
	fmt.Print(line)
}
```

</TabItem>
<TabItem label="Java" icon="seti:java">

```java
import io.daytona.sdk.Daytona;
import io.daytona.sdk.Image;

public class App {
    public static void main(String[] args) {
        try (Daytona daytona = new Daytona()) {
            Image image = Image.base("daytonaio/sandbox")
                .runCommands(
                    "sudo apt-get update "
                        + "&& sudo apt-get install -y --no-install-recommends "
                        + "ca-certificates fuse3 libssl3 openssl "
                        + "&& sudo update-ca-certificates",
                    "curl -fsSL https://mesa.dev/install.sh | sh "
                        + "&& sudo ln -sf $HOME/.mesa/bin/mesa /usr/local/bin/mesa",
                    "sudo sed -i 's/^#user_allow_other/user_allow_other/' /etc/fuse.conf"
                );

>             daytona.snapshot().create("fuse-mesa", image, System.out::println);

Reviewer: @OliverGilan, I think this could be simpler. Can you please confirm:

1. I only believe we need the `user_allow_other` change
2. I think this is only relevant `curl -fsSL https://mesa.dev/install.sh | sh `
3. The `ln` feels like an awful hack -- symlinking a user-installed binary into `/usr/local/bin` is a bad idea in general, for a multitude of reasons, and I believe that editing `PATH` is better. Perhaps we could make the `install.sh` script on our end support root installs.

        }
    }
}
```

</TabItem>
</Tabs>

#### Launch and mount

> Pass `MESA_API_KEY` and your Mesa organization slug to the sandbox via `envVars`. The sandbox writes a TOML config, authenticates the Mesa CLI, and mounts your repos at `/home/daytona/mesa/mnt/<org>/<repo>`.

Reviewer: ```suggestion
Pass `MESA_API_KEY` and your Mesa organization slug to the sandbox via `envVars`. Your code then writes a TOML config into the sandbox, authenticates the Mesa CLI, and mounts your repos at `/home/daytona/mesa/mnt/<org>/<repo>`.
```

        env_vars={
            "MESA_API_KEY": ephemeral_key,
            "MESA_ORG": os.environ["MESA_ORG"],
        },
    )
)
```

</TabItem>
<TabItem label="TypeScript" icon="seti:typescript">

```typescript
import { Daytona } from '@daytona/sdk'
import { Mesa } from '@mesadev/sdk'

const mesa = new Mesa({ apiKey: process.env.MESA_API_KEY!, org: process.env.MESA_ORG! })

const ephemeralKey = await mesa.apiKeys.create({
  name: 'sandbox-session',
  scopes: ['read', 'write'],
  expires_in_seconds: 3600,
})

const daytona = new Daytona()
const sandbox = await daytona.create({
  snapshot: 'fuse-mesa',
  envVars: {
    MESA_API_KEY: ephemeralKey.key,
    MESA_ORG: process.env.MESA_ORG!,
  },
})
```

</TabItem>
</Tabs>

The rest of the launch flow (writing the TOML config, `mesa auth set-key`, `mesa mount --daemonize`) is unchanged — the sandbox doesn't know whether the key it received is long-lived or ephemeral.

For repo-scoped or path-scoped keys, see Mesa's [auth and permissions guide ↗](https://docs.mesa.dev/content/getting-started/auth-and-permissions). For the full integration recipe, see Mesa's [Daytona guide ↗](https://docs.mesa.dev/content/integration-guides/daytona).

> ## Unmounting

Reviewer 2: ```suggestion
## Unmount
```

> When a sandbox is deleted via `daytona.delete(sandbox)`, the container teardown automatically removes any active FUSE mounts and shuts down their daemons. For normal cleanup this is all you need — no manual unmount required.

Reviewer: ```suggestion
When a sandbox is deleted via `daytona.delete(sandbox)`, the container teardown automatically removes any active FUSE mounts and shuts down their daemons. For normal cleanup, this is all you need — no manual unmount required.
```

To free a mount path **during** a sandbox's lifetime (for example, to remount with different credentials or before persisting a workspace archive), relocate the mount onto a throwaway path:

```bash
sudo mkdir -p /tmp/.fuse-defunct-$$
sudo mount --move <your-mount-path> /tmp/.fuse-defunct-$$
```

> After this, your original mount path is free for re-mounting. The FUSE daemon stays alive serving the mount at the new path; both the relocated mount and the daemon are cleaned up automatically when the sandbox is deleted.

Reviewer: ```suggestion
After this, your original mount path is free for remounting. The FUSE daemon stays alive serving the mount at the new path; both the relocated mount and the daemon are cleaned up automatically when the sandbox is deleted.
```

> This works for any FUSE-based mount; verified against `mount-s3`, `gcsfuse`, and `blobfuse2`.

Reviewer: ```suggestion
This works for any FUSE-based mount — verified against `mount-s3`, `gcsfuse`, and `blobfuse2`.
```

## Tips

- **Snapshot versioning** — name snapshots like `fuse-s3-v1`. Bump the suffix to force a rebuild whenever you upgrade the FUSE tool.

> - **Credential security** — pass credentials via `env_vars` / `envVars` on sandbox creation rather than committing them into the snapshot. Env vars stay scoped to the sandbox lifetime and rotate without rebuilding the image.

Reviewer 2: I'd remove this as it would be explained as part of the changes above with the environment variables for each storage provider.

Author: Removed credentials bullet point, kept just the snapshot versioning one

## See also

- [Daytona Volumes](/docs/en/volumes/) — first-party FUSE-backed shared storage hosted by Daytona
- [Declarative Builder](/docs/en/declarative-builder/) — programmatic image and snapshot definitions
- [Snapshots](/docs/en/snapshots/) — pre-built sandbox templates

> - [File system operations](/docs/en/file-system-operations/) — uploading and downloading files in the sandbox

Reviewer 2: ```suggestion

```

External storage mounts and [Daytona Volumes](/docs/en/volumes/) are complementary FUSE-based mechanisms — both expose remote object storage as a regular sandbox directory, both can be shared across sandboxes, and both persist beyond any individual sandbox's lifetime. The main distinction is **where the data physically lives**: Daytona Volumes are hosted on Daytona's own S3-compatible object store, while external mounts connect to a bucket or filesystem hosted on another provider (Amazon S3, Cloudflare R2, GCS, Azure Blob, MesaFS).

External storage is mounted using FUSE. Daytona supports two approaches, and each provider section below shows both — pick whichever fits your workflow:

- **Pre-baked snapshot** — build a [snapshot](/docs/en/snapshots/) once with the FUSE tool (`mount-s3`, `gcsfuse`, `blobfuse2`) baked in, then launch every sandbox from that snapshot. Cold starts are fast and predictable. Best for production.

> - **Runtime install** — launch a default sandbox and `apt-get install` the FUSE tool when the sandbox starts. Adds time to sandbox startup, but you don't manage snapshots. Best for quick experiments.

Reviewer 2: ```suggestion
- **Pre-built snapshot** — build a [snapshot](/docs/en/snapshots/) once with the FUSE tool (`mount-s3`, `gcsfuse`, `blobfuse2`) built-in, then launch every sandbox from that snapshot. Cold starts are fast and predictable. Best for production.
- **Runtime install** — launch a default sandbox and `apt-get install` the FUSE tool when the sandbox starts. Adds time to sandbox startup, but you don't manage snapshots. Best for quick experiments.
```

Both approaches end with the same mount command and the same usage — the only difference is when the FUSE tool gets installed.

## Mount an Amazon S3 bucket

Mount an S3 bucket using [Mountpoint for Amazon S3 ↗](https://github.com/awslabs/mountpoint-s3) — AWS's official FUSE client, optimized for high throughput on S3.

**Credentials** — set `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY` in your local environment. The snippets below pass them into the sandbox via `envVars`, and `mount-s3` reads them from there.

### Pre-baked snapshot

> Build a snapshot once with `mount-s3` installed, then launch sandboxes from it.

Reviewer 2: ```suggestion
### Pre-built snapshot

Build a snapshot with `mount-s3` preinstalled, then launch all S3-enabled sandboxes from that snapshot. This removes per-sandbox package install work, keeps cold starts predictable, and gives you a reusable baseline image for production workloads.
```

#### Build a snapshot

> 

Reviewer 2: ```suggestion

Create a reusable snapshot that installs `mount-s3` and its system dependencies. After it finishes, every sandbox launched from `fuse-s3` already has the mount binary available.

```

        try (Daytona daytona = new Daytona()) {
            CreateSandboxFromSnapshotParams params = new CreateSandboxFromSnapshotParams();
            params.setSnapshot("fuse-s3");
            params.setEnvVars(Map.of(
                "AWS_ACCESS_KEY_ID", System.getenv("AWS_ACCESS_KEY_ID"),
                "AWS_SECRET_ACCESS_KEY", System.getenv("AWS_SECRET_ACCESS_KEY")
            ));
            Sandbox sandbox = daytona.create(params);

            String mountPath = "/home/daytona/s3";

            // mount-s3 daemonizes by default and reads AWS_* from the environment
            sandbox.getProcess().executeCommand("mkdir -p " + mountPath);
            sandbox.getProcess().executeCommand("mount-s3 my-bucket " + mountPath);

            // Read and write through the mount as if it were a local directory
            sandbox.getProcess().executeCommand(
                "echo 'hello from Daytona' > " + mountPath + "/hello.txt");
            ExecuteResponse response = sandbox.getProcess().executeCommand(
                "cat " + mountPath + "/hello.txt");
            System.out.println(response.getResult());
        }
    }
}
```

</TabItem>
</Tabs>

### Runtime install

> Skip the snapshot and install `mount-s3` when the sandbox starts. Same end result, slower cold start.

Reviewer 2: ```suggestion
Start from a default sandbox and install `mount-s3` during startup before running the mount command. This is useful for quick testing and temporary environments where you do not want to maintain a custom snapshot, with the tradeoff of slower cold starts.
```

                "arch=\"$(dpkg --print-architecture | sed s/amd64/x86_64/)\" "
                    + "&& wget -O /tmp/mount-s3.deb "
                    + "\"https://s3.amazonaws.com/mountpoint-s3-release/latest/${arch}/mount-s3.deb\" "
                    + "&& sudo apt-get install -y /tmp/mount-s3.deb");

            // Mount and use
            String mountPath = "/home/daytona/s3";
            sandbox.getProcess().executeCommand(
                "mkdir -p " + mountPath + " && mount-s3 my-bucket " + mountPath);
            ExecuteResponse response = sandbox.getProcess().executeCommand("ls " + mountPath);
            System.out.println(response.getResult());
        }
    }
}
```

</TabItem>
</Tabs>

## Mount a Cloudflare R2 bucket

Cloudflare R2 is S3-compatible, so the same `mount-s3` tool works. Pass an explicit `--endpoint-url` pointing at your R2 account.

**Credentials** — set `R2_ACCOUNT_ID`, `R2_ACCESS_KEY_ID`, and `R2_SECRET_ACCESS_KEY` in your local environment. R2 is S3-compatible, so the snippets below pass your R2 keys into the sandbox via `envVars` under the `AWS_ACCESS_KEY_ID` / `AWS_SECRET_ACCESS_KEY` names that `mount-s3` expects.

### Pre-baked snapshot

> Build a snapshot once with `mount-s3` installed (the same FUSE tool used for S3), then launch sandboxes from it.

Reviewer 2: ```suggestion
### Pre-built snapshot

Build a snapshot with `mount-s3` preinstalled, then launch all R2-enabled sandboxes from that snapshot. The mount flow stays identical to S3 except for the R2 `--endpoint-url`, and startup remains fast because installation is done once at snapshot build time.
```

#### Build a snapshot

> 

Reviewer 2: ```suggestion

Create a reusable snapshot that installs the same `mount-s3` tool used for S3. R2 remains S3-compatible, so this snapshot is identical to S3 setup and only the runtime mount command changes.

```

	Name:  "fuse-r2",
	Image: image,
})
if err != nil {
	log.Fatal(err)
}
for line := range logChan {
	fmt.Print(line)
}
```

</TabItem>
<TabItem label="Java" icon="seti:java">

```java
import io.daytona.sdk.Daytona;
import io.daytona.sdk.Image;

public class App {
    public static void main(String[] args) {
        try (Daytona daytona = new Daytona()) {
            Image image = Image.base("daytonaio/sandbox")
                .runCommands(
                    "sudo apt-get update "
                        + "&& sudo apt-get install -y --no-install-recommends libfuse2 ca-certificates wget",
                    "arch=\"$(dpkg --print-architecture | sed s/amd64/x86_64/)\" "
                        + "&& wget -O /tmp/mount-s3.deb "
                        + "\"https://s3.amazonaws.com/mountpoint-s3-release/latest/${arch}/mount-s3.deb\" "
                        + "&& sudo apt-get install -y /tmp/mount-s3.deb "
                        + "&& rm /tmp/mount-s3.deb"
                );

            daytona.snapshot().create("fuse-r2", image, System.out::println);
        }
    }
}
```

</TabItem>
</Tabs>

#### Launch and mount

> 

Reviewer 2: ```suggestion

Pass your R2 credentials into the sandbox as `AWS_*` environment variables and mount with the R2 endpoint URL. This keeps the authentication flow compatible with `mount-s3` while targeting your Cloudflare account.

```

public class App {
    public static void main(String[] args) {
        try (Daytona daytona = new Daytona()) {
            // R2 credentials live in your Cloudflare dashboard under R2 > Manage API Tokens
            String accountId = System.getenv("R2_ACCOUNT_ID");

            CreateSandboxFromSnapshotParams params = new CreateSandboxFromSnapshotParams();
            params.setSnapshot("fuse-r2");
            params.setEnvVars(Map.of(
                "AWS_ACCESS_KEY_ID", System.getenv("R2_ACCESS_KEY_ID"),
                "AWS_SECRET_ACCESS_KEY", System.getenv("R2_SECRET_ACCESS_KEY")
            ));
            Sandbox sandbox = daytona.create(params);

            String mountPath = "/home/daytona/r2";

            sandbox.getProcess().executeCommand("mkdir -p " + mountPath);
            sandbox.getProcess().executeCommand(
                "mount-s3 --endpoint-url https://" + accountId + ".r2.cloudflarestorage.com "
                    + "my-r2-bucket " + mountPath);

            ExecuteResponse response = sandbox.getProcess().executeCommand("ls " + mountPath);
            System.out.println(response.getResult());
        }
    }
}
```

</TabItem>
</Tabs>

### Runtime install

> 

Reviewer 2: ```suggestion

Start from a default sandbox and install `mount-s3` during startup, then mount your bucket with the R2 `--endpoint-url`. This path is convenient for prototyping or one-off tasks, but each new sandbox pays the package installation cost.

```

                    + "&& sudo apt-get install -y /tmp/mount-s3.deb");

            // Mount with R2 endpoint
            String mountPath = "/home/daytona/r2";
            sandbox.getProcess().executeCommand(
                "mkdir -p " + mountPath + " && "
                    + "mount-s3 --endpoint-url https://" + accountId + ".r2.cloudflarestorage.com "
                    + "my-r2-bucket " + mountPath);

            ExecuteResponse response = sandbox.getProcess().executeCommand("ls " + mountPath);
            System.out.println(response.getResult());
        }
    }
}
```

</TabItem>
</Tabs>

## Mount a Google Cloud Storage bucket

Mount a GCS bucket using [gcsfuse ↗](https://github.com/GoogleCloudPlatform/gcsfuse) — Google's official FUSE client.

**Credentials** — `gcsfuse` reads a service account JSON key file. The snippets below read the key from a local path on your host and upload it into the sandbox via `sandbox.fs`.

:::note
Daytona's base image is Debian Trixie, but Google has not published a Trixie-specific gcsfuse repository yet. The `gcsfuse-bookworm` repo packages run cleanly on Trixie, so we use them in the snippets below.
:::

### Pre-baked snapshot

> 

Reviewer 2: ```suggestion
### Pre-built snapshot

Build a snapshot with `gcsfuse` preinstalled, then launch all GCS-enabled sandboxes from that snapshot. This avoids repeating apt repository setup and package installation for every sandbox, which makes startup behavior more consistent.

```

> #### Build a snapshot

Reviewer 2: ```suggestion
#### Build a snapshot

Create a reusable snapshot that installs `gcsfuse` plus its apt repository configuration. After this step, GCS-enabled sandboxes can mount immediately without repeating package setup.

```

            // GCS_SERVICE_ACCOUNT_KEY holds the full service account JSON as a string
            byte[] serviceAccountKey = System.getenv("GCS_SERVICE_ACCOUNT_KEY")
                .getBytes(StandardCharsets.UTF_8);

            CreateSandboxFromSnapshotParams params = new CreateSandboxFromSnapshotParams();
            params.setSnapshot("fuse-gcs");
            Sandbox sandbox = daytona.create(params);

            String mountPath = "/home/daytona/gcs";
            String keyPath = "/home/daytona/.gcs-key.json";

            // Upload the key file into the sandbox
            sandbox.fs.uploadFile(serviceAccountKey, keyPath);
            sandbox.getProcess().executeCommand("chmod 600 " + keyPath);

            // Mount the bucket
            sandbox.getProcess().executeCommand("mkdir -p " + mountPath);
            sandbox.getProcess().executeCommand(
                "gcsfuse --key-file=" + keyPath + " my-gcs-bucket " + mountPath);

            // Use the mount
            ExecuteResponse response = sandbox.getProcess().executeCommand("ls " + mountPath);
            System.out.println(response.getResult());
        }
    }
}
```

</TabItem>
</Tabs>

> ### Runtime install

Reviewer 2: ```suggestion
### Runtime install

Start from a default sandbox and install `gcsfuse` when the sandbox starts, then upload the service account key and mount the bucket. This is the fastest way to iterate on setup, but every sandbox repeats install and key staging steps.

```

            sandbox.getProcess().executeCommand("mkdir -p " + mountPath);
            sandbox.getProcess().executeCommand(
                "gcsfuse --key-file=" + keyPath + " my-gcs-bucket " + mountPath);

            ExecuteResponse response = sandbox.getProcess().executeCommand("ls " + mountPath);
            System.out.println(response.getResult());
        }
    }
}
```

</TabItem>
</Tabs>

## Mount an Azure Blob container

Mount an Azure Blob container using [blobfuse2 ↗](https://github.com/Azure/azure-storage-fuse) — Microsoft's official FUSE client.

**Credentials** — set `AZURE_STORAGE_ACCOUNT`, `AZURE_STORAGE_CONTAINER`, and `AZURE_STORAGE_ACCOUNT_KEY` in your local environment. The snippets below pass them into the sandbox via `envVars`, and `blobfuse2` reads them from its YAML config.

:::caution
Daytona's base image is Debian Trixie. Microsoft's `blobfuse2` Bookworm binary links against `libfuse3.so.3`, but Trixie's `fuse3` package bumped the SONAME to `.4`, so `libfuse3.so.3` is missing on disk. The snapshot recipe below creates a compatibility symlink. Without it, `blobfuse2` fails with `libfuse3.so.3: cannot open shared object file`.
:::

> ### Pre-baked snapshot

Reviewer 2: ```suggestion
### Pre-built snapshot

Build a snapshot with `blobfuse2` and required FUSE compatibility setup preinstalled, then launch all Azure-enabled sandboxes from that snapshot. This is the recommended path for stable environments because dependency and compatibility work runs once during snapshot creation.

```

> #### Build a snapshot

Reviewer 2: ```suggestion
#### Build a snapshot

Create a reusable snapshot that installs `blobfuse2`, configures required FUSE dependencies, and applies the Trixie compatibility steps. This ensures Azure mounts work out of the box in sandboxes launched from `fuse-azure`.

```

                + "components:\n"
                + "  - libfuse\n"
                + "  - block_cache\n"
                + "  - attr_cache\n"
                + "  - azstorage\n"
                + "azstorage:\n"
                + "  type: block\n"
                + "  account-name: " + account + "\n"
                + "  container: " + container + "\n"
                + "  endpoint: https://" + account + ".blob.core.windows.net\n"
                + "  auth-type: key\n"
                + "  account-key: " + accountKey + "\n";

            sandbox.fs.uploadFile(config.getBytes(StandardCharsets.UTF_8), configPath);
            sandbox.getProcess().executeCommand("chmod 600 " + configPath);

            // Mount the container
            sandbox.getProcess().executeCommand("mkdir -p " + mountPath);
            sandbox.getProcess().executeCommand(
                "blobfuse2 mount --config-file=" + configPath + " " + mountPath);

            // Use the mount
            ExecuteResponse response = sandbox.getProcess().executeCommand("ls " + mountPath);
            System.out.println(response.getResult());
        }
    }
}
```

</TabItem>
</Tabs>

> ### Runtime install

Reviewer 2: ```suggestion
### Runtime install

Start from a default sandbox and install `blobfuse2` during startup before writing the config and mounting the container. This is useful for quick validation and experiments, with the tradeoff of slower cold starts and repeated setup on each sandbox launch.

```

            ExecuteResponse response = sandbox.getProcess().executeCommand("ls " + mountPath);
            System.out.println(response.getResult());
        }
    }
}
```

</TabItem>
</Tabs>

## Mount a MesaFS filesystem

[MesaFS ↗](https://mesa.dev) is an agent-native versioned filesystem from Mesa, purpose-built for the same workloads Daytona sandboxes run — parallel agent swarms, shared working memory, structured artifacts, and long-lived state across runs. With MesaFS, instead of mounting a cloud bucket, you mount a Mesa **repository**: a Git-compatible versioned working directory with sub-50ms reads/writes, instant fork/branch operations, and unlimited concurrent writers.

The Mesa setup follows the same pattern as the bucket providers but uses the Mesa CLI rather than a FUSE-specific tool: install the CLI in your sandbox, authenticate with your API key, and run `mesa mount --daemonize` to mount your repos at `/home/daytona/mesa/mnt/<org>/<repo>`.

**Credentials** — set `MESA_API_KEY` and `MESA_ORG` (your Mesa organization slug) in your local environment. The snippets below pass them into the sandbox via `envVars`, and the Mesa CLI reads them from there.

> ### Pre-baked snapshot

Reviewer 2: ```suggestion
### Pre-built snapshot

Build a snapshot with the Mesa CLI preinstalled, then launch Mesa-enabled sandboxes from that snapshot. You still authenticate and mount at runtime, but installation is no longer part of each sandbox startup sequence.

```

> #### Build a snapshot

Reviewer 2: ```suggestion
#### Build a snapshot

Create a reusable snapshot that installs the Mesa CLI and enables the FUSE `user_allow_other` setting. Sandboxes launched from `fuse-mesa` can then authenticate and mount repos without repeating install work.

```

            CreateSandboxFromSnapshotParams params = new CreateSandboxFromSnapshotParams();
            params.setSnapshot("fuse-mesa");
            params.setEnvVars(Map.of(
                "MESA_API_KEY", System.getenv("MESA_API_KEY"),
                "MESA_ORG", org
            ));
            Sandbox sandbox = daytona.create(params);

            String config = "mount-point = \"/home/daytona/mesa/mnt\"\n\n"
                + "[secrets]\n"
                + "backend = \"plaintext-file\"\n\n"
                + "[organizations." + org + "]\n";

            sandbox.getProcess().executeCommand("mkdir -p $(dirname " + configPath + ")");
            sandbox.fs.uploadFile(config.getBytes(StandardCharsets.UTF_8), configPath);

            sandbox.getProcess().executeCommand(
                "mesa auth set-key --org $MESA_ORG $MESA_API_KEY");
            sandbox.getProcess().executeCommand("mesa mount --daemonize");

            ExecuteResponse response = sandbox.getProcess().executeCommand("ls " + mountPath);
            System.out.println(response.getResult());
        }
    }
}
```

</TabItem>
</Tabs>

### Runtime install

> Skip the snapshot and install the Mesa CLI when the sandbox starts. Same end result, slower cold start.

Reviewer 2: ```suggestion
Start from a default sandbox and install the Mesa CLI during startup before configuring auth and running `mesa mount --daemonize`. This is useful when iterating quickly on mount behavior, with the tradeoff of slower cold starts for each sandbox.
```

For repo-scoped or path-scoped keys, see Mesa's [auth and permissions guide ↗](https://docs.mesa.dev/content/getting-started/auth-and-permissions). For the full integration recipe, see Mesa's [Daytona guide ↗](https://docs.mesa.dev/content/integration-guides/daytona).

## Unmount

When a sandbox is deleted via `daytona.delete(sandbox)`, the container teardown automatically removes any active FUSE mounts and shuts down their daemons. For normal cleanup, this is all you need — no manual unmount required.

To free a mount path **during** a sandbox's lifetime (for example, to remount with different credentials or before persisting a workspace archive), relocate the mount onto a throwaway path:

```bash
sudo mkdir -p /tmp/.fuse-defunct-$$
sudo mount --move <your-mount-path> /tmp/.fuse-defunct-$$
```

After this, your original mount path is free for remounting. The FUSE daemon stays alive serving the mount at the new path; both the relocated mount and the daemon are cleaned up automatically when the sandbox is deleted.

This works for any FUSE-based mount — verified against `mount-s3`, `gcsfuse`, and `blobfuse2`.

## Tips

> - **Snapshot versioning** — name snapshots like `fuse-s3-v1`. Bump the suffix to force a rebuild whenever you upgrade the FUSE tool.

Reviewer 2: We could also move this above in the introduction section or put somewhere as part of a section so it's not standalone at the end of the file.

Author: I deleted it since the tip idea is pretty straightforward and intuitive. Also it might change in future if we implement the latest tag feature
